Added configurable subject and content for Flowdock messages

timothydowney opened the pull request:
I've added configurable content and subject options for Flowdock chat & team inbox messages.  I've also tidied up the DataBoundConstructor with DataBoundSetters to better reflect the optional parameters.  This in turn makes the generate job DSL a bit cleaner.  Here's an example usage:

```
                    flowdockNotifier {
                        flowToken(devOpsFlowToken)
                        notificationTags("docker-dashboard, ${JOB_NAME}")
                        subject("[${k}] Deployment for ${project.projectName} of \$VERSION")
                        content("[${k}] Deployment for ${project.projectName} of \$VERSION by \$BUILD_USER_EMAIL run from ${v.deployServer}")
                    }
```

timothydowney committed:
Added configurable subject and content plus some small refactors to improve the auto generated DSL for job dsl

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -14,54 +25,34 @@
+    private boolean chatNotification;
 
     private final Map<BuildResult, Boolean> notifyMap;
-    private final boolean notifySuccess;
```
I do not see the migration code for the Old Data from previous versions

timothydowney replied:
Hi Oleg -- Thanks for taking a look so quickly.  I assumed (incorrectly, I guess) that since the values were stored in XML as strings that they would be stapled properly in.  I was basically trying to clean things up so that proper types could be used (boolean vs string) for the different options.  That in turn simplified a lot of the optional parameters by moving them out of the constructor into DataBoundSetters.

Do you have an example of what I can look at to ensure that the old data from previous versions migrates successfully?

oleg-nenashev replied:
Here is the guide: https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -14,54 +25,34 @@
-    public FlowdockNotifier(String flowToken, String notificationTags, String chatNotification,
-        String notifySuccess, String notifyFailure, String notifyFixed, String notifyUnstable,
-        String notifyAborted, String notifyNotBuilt) {
+    public FlowdockNotifier(String flowToken) {
```
Formally it breaks the binary compatibility. Not sure if anybody cares in the case of this plugin, which is not an API one

timothydowney replied:
Thanks again....I wasn't sure if compatibility would be broken by this since the parameter names are the same.  I guess that I assumed incorrectly that the stapler would take care of this.  All of the parameters are still there, they are just now DataBoundSetters.  I did this because most of the constructor parameters are for optional arguments.

I made this change so as to be able to use a proper type for the booleans.  Is there a way to make this type of change and still preserve compatibility?  I was hoping to be able to make the change since it makes the generate job DSL cleaner and more what you'd expect when dealing with boolean fields.

If there's no way to do it, I guess I'll have to revert that part of the change.

oleg-nenashev replied:
> I made this change so as to be able to use a proper type for the booleans. Is there a way to make this type of change and still preserve compatibility? 

Just create another public constructor, which retains the original interface. Then, to not set `DataBoundConstructor` there, but add `@Deprecated` and `@Restricted(NoExternalUse.class)`

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -77,22 +68,22 @@ public boolean getChatNotification() {
     }
 
     public boolean getNotifySuccess() {
```
All the code below will throw `NullPointerException` if the BuildResult is missing in the list.
I would suggest creating a utility method, which handles such case

timothydowney replied:
Given that I've changed to initialize the list with all possible values of the enum, I think that this is safe now.

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -119,18 +110,44 @@ public boolean shouldNotify(BuildResult buildResult) {
+     * @return True if the value is non-null and has at least one non-blank character.
+     */
+    private boolean isValueSet(String value) {
+    	return value != null && value.trim().length() > 0;
```
`StringUtils.isNotBlank()` is a full replacement of this method

timothydowney replied:
Updated based on your suggestion.

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -198,4 +215,62 @@ public String apiUrl() {
         }
     }
+
+	public String getSubject() {
```
Tabs vs. spaces in indentation

timothydowney replied:
Should be fixed now.  Again, my apologies for the whitespace damages to the diffs.

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -198,4 +215,62 @@ public String apiUrl() {
+
+	@DataBoundSetter
+	public void setContent(String content) {
+		this.content = content;
```
I would explicitly make the value `null` if the input string is blank

timothydowney replied:
Done.

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -198,4 +215,62 @@ public String apiUrl() {
+
+	@DataBoundSetter
+	public void setSubject(String subject) {
+		this.subject = subject;
```
I would explicitly make the value `null` if the input string is blank

timothydowney replied:
Done.

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -198,4 +215,62 @@ public String apiUrl() {
+		return subject;
+	}
+
+	public String getContent() {
```
Annotations like `@CheckForNull` could prevent misusages in API

timothydowney replied:
Added based on your suggestion.

oleg-nenashev commented on src/main/java/com/flowdock/jenkins/FlowdockNotifier.java:
```diff
@@ -14,54 +25,34 @@
+        this.notifyMap.put(BuildResult.FIXED, true);
+        this.notifyMap.put(BuildResult.UNSTABLE, true);
+        this.notifyMap.put(BuildResult.ABORTED, true);
+        this.notifyMap.put(BuildResult.NOT_BUILT, true);
```
It is easy to forget setting something when you extend the enum. Maybe it's better to support a default behavior for missing entries

timothydowney replied:
This is now handled that all values of the enum are initialized  to a default of true.  Any new values added would be covered via the loop.  This assumes that true is a good default, which has been the case of the pre-existing behaviors.

oleg-nenashev reviewed (changes requested):
There is a problem with data migration and missing cache entries, other comments are rather not important

timothydowney committed:
Addressing code review for PR-27

timothydowney commented:
Hi Oleg -- I think that I've addressed your requests.  I've added defaulting for the enum across the board, restored the legacy constructor and marked it deprecated, plus also added the style changes you suggested.  Thanks a lot for the review.

The one other question I had was on retaining backward compatibility on the legacy changes.  I had changed the type on existing fields, but the names are the same and added two new fields.  The fields get initialize via the constructor to the proper defaults.  Is that enough to maintain compatibility or is there more?  It didn't seem like readResolve would be needed for simple types, but I wasn't sure if I was misreading.

Thanks!
Tim

timothydowney commented:
Oh...one more thing.  I want to write some unit tests for this thing.  I've written unit tests for pipeline based plugins but not for these older style plugins.  Have you seen a good page on the wiki or do you know of a good representative plugin that I should take a look at for comparison?

timothydowney commented:
Never mind...I should have searched first.  Found this.

https://wiki.jenkins-ci.org/display/JENKINS/Unit+Test

I'll write some tests.

timothydowney commented:
Hi Oleg -- I'll be updating the PR tomorrow with unit tests.  I've got most of this thing tested right now and was hoping to finish this afternoon, but regular work caught up with me.  I'll update the PR once I'm done with the unit tests.

timothydowney committed:
Added variety of unit tests

timothydowney commented:
Ok -- I've added several unit tests.  Had to do some minor refactors to make things more testable with Mockito.  This puts things in a better spot for some refactors in the future that may be necessary to make this more compatible with pipeline.

Please let me know if you have any other suggestions.

oleg-nenashev commented:
Hard to review the diffs due to spaces vs. tabs. The original plugin codebase uses spaces, so I would suggest sticking to it. Some of previous comments have been hidden due to indenting changes from what I see

timothydowney committed:
Fixing whitespace issue -- Should be spaces

timothydowney committed:
Revert "Fixing whitespace issue -- Should be spaces"

This reverts commit 7e4f5c97c0aaa7bbd976742b52c7dcf590b3a40a.

timothydowney committed:
Still trying to fix the whitespace changes

timothydowney committed:
Missed a few more accidental whitespace changes

timothydowney commented:
I apologize for the whitespace changes.  My IDE claimed that it was set up to use space instead of tabs, but it was not.  I've gone back and attempted to clean it all up as much as possible.  There are still some situations where the imports were reorganized by the IDE that have caused a few extra diffs.  Hopefuly it's ok.

I've ended up with a few extra commits here trying to clean this whole mess up.  Sorry about that.  I looked through the diffs here via github and they seem readable enough now.

Thanks for your patience and please let me know if you have any other suggestions.

timothydowney commented:
Hi @oleg-nenashev , Were you able to read the diffs well enough?  Sorry again about the whitespace challenges.  If they're still difficult, I could redo the changes in a fresh PR.  Please let me know what is easiest for you.  Thanks again.

mletterle commented:
These unit tests would be very handy to have merged in... I'm considering starting to work on pipeline compatibility and they would help a lot...
